erc721/ERC721.vy

```diff
@@ -249,64 +245,76 @@ def safeTransferFrom(
         _tokenId: uint256,
         _data: bytes[1024]=""
     ):
-    self._validateTransferFrom(_from, _to, _tokenId, msg.sender)
-    self._doTransfer(_from, _to, _tokenId)
-    if(_to.codesize > 0): # check if the _to is a contract address
+    self._transferFrom(_from, _to, _tokenId, msg.sender)
+    if(_to.codesize > 0): # check if `_to` is a contract address
         returnValue: bytes32 = ERC721Receiver(_to).onERC721Received(msg.sender, _from, _tokenId, _data)
         assert returnValue == method_id("onERC721Received(address,address,uint256,bytes)", bytes32)
 
 
-# @dev Set or reaffirm the approved address for an NFT.
-# @notice The zero address indicates there is no approved address. Throws unless `msg.sender` is
-#         the current NFT owner, or an authorized operator of the current owner.
+# @dev Set or reaffirm the approved address for an NFT. The zero address indicates there is no approved address. 
+#      Throws unless `msg.sender` is the current NFT owner, or an authorized operator of the current owner.
+#      Throws if `_tokenId` is not a valid NFT. (NOTE: This is not written the EIP)
+#      Throws if `_approved` is the current owner. (NOTE: This is not written the EIP)
 # @param _approved Address to be approved for the given NFT ID.
 # @param _tokenId ID of the token to be approved.
 @public
 def approve(_approved: address, _tokenId: uint256):
-    # get owner
-    owner: address = self._ownerOf(_tokenId)
+    owner: address = self.idToOwner[_tokenId]
+    # Throws if `_tokenId` is not a valid NFT
+    assert owner != ZERO_ADDRESS
+    # Throws if `_approved` is the current owner
     assert _approved != owner
-    # check requirements
+    # Check requirements
     senderIsOwner: bool = self.idToOwner[_tokenId] == msg.sender
-    assert (senderIsOwner or self._isApprovedForAll(owner, msg.sender))
-    # set the approval
+    senderIsApprovedForAll: bool = (self.ownerToOperators[owner])[msg.sender]
+    assert (senderIsOwner or senderIsApprovedForAll)
+    # Set the approval
     self.idToApprovals[_tokenId] = _approved
     log.Approval(owner, _approved, _tokenId)
 
 
 # @dev Enables or disables approval for a third party ("operator") to manage all of
 #      `msg.sender`'s assets. It also emits the ApprovalForAll event.
-# @notice This works even if sender doesn't own any tokens at the time.
+#      NOTE: This works even if sender doesn't own any tokens at the time.
 # @param _operator Address to add to the set of authorized operators.
 # @param _isApproved True if the operators is approved, false to revoke approval.
 @public
 def setApprovalForAll(_operator: address, _approved: bool):
```

Comment:
[IMO]
It is better to check `_operator != owner` at the top of `setApprovalForAll`

Reply:
hmm... I think it's OK to approve for themselves?

Reply:
Yes it is basically ok, but when executed (by mistake), more gas is consumed.
To approve themselves is needed?

Reply:
OK, to prevent `approve` and `setApprovalForAll` to zero address and sender are not required in the original EIP but improve security or avoid cases the sender probably don't intend. Whether or not to add these assertions is different among implementations and it has a small trade-off of gas cost.

Reply:
agree

Reply:
fixed